packages/shield-controller/src/ShieldController.ts

```diff
@@ -0,0 +1,259 @@
+ */
+export type ShieldControllerActions = ShieldControllerCheckCoverageAction;
+
+export type ShieldControllerCoverageResultReceivedEvent = {
+  type: `${typeof controllerName}:coverageResultReceived`;
+  payload: [coverageResult: CoverageResult];
+};
+
+export type ShieldControllerStateChangeEvent = ControllerStateChangeEvent<
+  typeof controllerName,
+  ShieldControllerState
+>;
+
+/**
+ * The internal events available to the ShieldController.
+ */
+export type ShieldControllerEvents =
+  | ShieldControllerCoverageResultReceivedEvent
+  | ShieldControllerStateChangeEvent;
+
+/**
+ * The external actions available to the ShieldController.
+ */
+export type AllowedActions = never;
+
+/**
+ * The external events available to the ShieldController.
+ */
+export type AllowedEvents = TransactionControllerStateChangeEvent;
+
+/**
+ * The messenger of the {@link ShieldController}.
+ */
+export type ShieldControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  ShieldControllerActions | AllowedActions,
+  ShieldControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * Metadata for the ShieldController state, describing how to "anonymize"
+ * the state and which parts should be persisted.
+ */
+const metadata = {
+  coverageResults: {
+    persist: true,
+    anonymous: false,
+  },
+  orderedTransactionHistory: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+export type ShieldControllerOptions = {
+  messenger: ShieldControllerMessenger;
+  state?: Partial<ShieldControllerState>;
+  backend: ShieldBackend;
+  transactionHistoryLimit?: number;
+  coverageHistoryLimit?: number;
+};
+
+export class ShieldController extends BaseController<
+  typeof controllerName,
+  ShieldControllerState,
+  ShieldControllerMessenger
+> {
+  readonly #backend: ShieldBackend;
+
+  readonly #coverageHistoryLimit: number;
+
+  readonly #transactionHistoryLimit: number;
+
+  readonly #transactionControllerStateChangeHandler: (
+    transactions: TransactionMeta[],
+    previousTransactions: TransactionMeta[] | undefined,
+  ) => void;
+
+  constructor(options: ShieldControllerOptions) {
+    const {
+      messenger,
+      state,
+      backend,
+      transactionHistoryLimit = 100,
+      coverageHistoryLimit = 10,
+    } = options;
+    super({
+      name: controllerName,
+      metadata,
+      messenger,
+      state: {
+        ...getDefaultShieldControllerState(),
+        ...state,
+      },
+    });
+
+    this.#backend = backend;
+    this.#coverageHistoryLimit = coverageHistoryLimit;
+    this.#transactionHistoryLimit = transactionHistoryLimit;
+    this.#transactionControllerStateChangeHandler =
+      this.#handleTransactionControllerStateChange.bind(this);
+  }
+
+  start() {
+    this.messagingSystem.subscribe(
+      'TransactionController:stateChange',
+      this.#transactionControllerStateChangeHandler,
+      (state) => state.transactions,
```

Comment:
Good use of a selector function, nice!

Reply:
Was a recommendation by @matthewwalsh0  ;)